[RestEasy Reactive Client] Use AsyncInputStream for Posting InputStream

fredericBregier opened the pull request:
InputStream was not taking into account when using sending it with ReastEasy client to a Rest service.

Why:
The previous implementation was putting all bytes in one buffer, leading either to OOME or limit on acceptable Buffer size.

Change:
Add InputStream Async support such that it will not fill the memory and respect back pressure.

Note that, to work, quarkus.http.limits.max-body-size shall be set to a correct value (ex. 0), since currently VertxInputStream has soft limit on this value. The behavior is not changed here but could if necessary since it is no more in memory (not one buffer but chunked mode with back pressure). But as setting the value to 0 or enough is working. 

Add a test to check correctness (speed and memory) but this can be improved since it only checks at the end the status and/or the size of the inputstream.

fredericBregier committed:
Use AsyncInputStream for Posting InputStream

InputStream was not taking into account when using sending it with
ReastEasy client to a Rest service.

Why:
The previous implementation was putting all bytes in one buffer, leading
either to OOME or limit on acceptable Buffer size.

Change:
Add InputStream Async support such that it will not fill the memory and
respect back pressure.

fredericBregier committed:
Apply same optimization on all VertxInputStream

fredericBregier committed:
Revert limiting synchronization and Add maxChunkSize

maxChunkSize to current MaxChunkSize, at least 8192

fredericBregier commented:
Note 2: POST is about 4 times slower than GET, but still better than nothing. 
Note 3: Using Netty as low level client side, performances are 4 times better on GET, 50% better on POST. But that's another story.

cescoffier reviewed (commented):
Thanks! 
A quickl look didn't see anything bad, but we need to be careful.

We would need the review from @geoand and @vietj.

geoand commented:
Very nice. This will take some time to review, more that the few minutes I currently have on my phone 😎

fredericBregier committed:
Fix import order

fredericBregier commented:
Just a side note: when using direct netty client, out of vertx or quarkus, I used HttpChunkedInput with a chunkWriteHandler to get max performance, but I didn't find a way to use it within reactive client, even if under there is netty...

fredericBregier committed:
Speedup retrieving inputstream

Workd both on sending and retrieving (server, client) an InputStream

Fetch more than 1 at startup (3 to still have a limit of memory
consumption)

Speedup:
- POST: from 900 MB/s to 1200 MB/s (almost close to Netty native Post
  at 1400 MB/s)
- GET: from 900 MB/s to 2000 MB/s (still half og Netty netive GET at
  4400 MB/s)

fredericBregier commented:
A bit improvement by fetching 3 instead of 1 at startup (so little memory consumption) and fetching each time a buffer is consumed (this seems to enable better pipelining between consumption and network):

POST: from 950 MB/s to 1200 MB/s, compare to Netty as client only up to 1400 MB/s
GET: from 950 MB/s to 2000 MB/s, compare to  Netty as client only up to 4400 MB/s

Note that increasing the startup fetching from 2 to 5 still increases the result, with still no memory impact (not noticeable):
- POST : 1) 950 MB/s 2) 1100 MB/s  3) 1200 MB/s 4) 1250 MB/s 5) 1300 MB/s
- GET: 1) 950 MB/s 2) 1500 MB/s 3) 2000 MB/s 4) 2250 MB/s 5) 2300 MB/s

My conviction is that 3 shall be the conservative option to get the best average performance while minimizing the memory impact (even if not noticeable). 4 might be the best choice however. (maybe an option for that ?)

geoand commented:
@vietj can you have a look at this as it very intense on the Vert.x side?

Thanks

vietj commented on independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/AsyncInputStream.java:
```diff
@@ -0,0 +1,201 @@
+    // Based on the inputStream with the real data
+    private final InputStream in;
+    private final Context context;
+    private final InboundBuffer<Buffer> queue;
```
using this is _ok_ but it remains an internal construct of vertx that might change.

fredericBregier replied:
I agree. We could use a standard queue (or a concurrent one)

vietj commented:
@fredericBregier I had a quick look: as far as I understood the AsyncInputStream will always be using the VertxBlockingInput so we have the guarantee that using the input stream never blocks ?

fredericBregier commented:
@vietj This implementation is inspired from various projects (not mines).
Of course, I cannot be certain 100%. I tried several ways.  I have an issue sometimes, never in simple POST (one client -> one server), but when chaining (client -> (server -> internal client) -> another server). I tried to find out why, but didn't get the reason yet.

However, if there is another way, I can try if you have any idea.

The issue right now is that sending an inputstream (POST) from a client leads to all bytes in memory, which leads to OOME. On the reverse way, getting InputStream (GET) is working (modulo some optimizations that I proposed too, but can be split in 2 MR).

fredericBregier commented:
@vietj And indeed, only in double POST (client to server to another server), it is indeed VertxBlockingInput that seems to block. But as it is under the wood, it seems difficult to bypass it.
(Note again: in single POST, no chaining, no issue)

Any idea ?

fredericBregier commented:
Just to give you more details: 

I update the scenario to show the issue using a self hack to include a deployed version of this.

```
@RegisterRestClient
@Path(ROOT_NORMAL)
public interface ApiInterface  {

  @POST
  Uni<Response> simplePost(InputStream inputStream);
}

@RegisterRestClient
@Path(ROOT_SECOND)
public interface ApiSecondInterface  {

  @POST
  Uni<Response> simplePost(InputStream inputStream);
}

@ApplicationScoped
public class ApiClient {
  @RestClient
  ApiInterface apiInterface;

  public long sendInputStream(InputStream inputStream) {
    Response response = apiInterface.simplePost(inputStream).await().indefinitely();
    return Long.parseLong(response.getHeaderString(HEADER_MAX_SIZE));
  }
}

@ApplicationScoped
public class ApiSecondClient {
  @RestClient
  ApiSecondInterface apiInterface;

  public long sendInputStream(InputStream inputStream) {
    Response response = apiInterface.simplePost(inputStream).await().indefinitely();
    return Long.parseLong(response.getHeaderString(HEADER_MAX_SIZE));
  }
}

@ApplicationScoped
@Path(ROOT_NORMAL)
public class ApiServer implements ApiInterface {
  @Inject
  ApiSecondClient client;
  
  @Override
  @Blocking
  public Uni<Response> simplePost(final InputStream inputStream) {
    return Uni.createFrom().emitter(em -> {
      final Semaphore semaphore = new Semaphore(0);
      final AtomicLong atomicLong = new AtomicLong();
      new Thread() { // To not block IO Thread... but maybe a better way ? Maybe the reason ?
        public void run() {
          try {
            long max = client.sendInput(inputStream);
            atomicLong.set(max);
          } finally {
            semaphore.release();
          }
        }
      }.start();
      try {
        semaphore.acquire();
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
      em.complete(Response.ok().header(HEADER_MAX_SIZE, atomicLong.get()).build());
    });
  }
}

@ApplicationScoped
@Path(ROOT_SECOND)
public class ApiSecondServer implements ApiSecondInterface {
  @Override
  @Blocking
  public Uni<Response> simplePost(final InputStream inputStream) {
    return Uni.createFrom().emitter(em -> {
      long max = ReadAll.readAll(inputStream); // simple method that read everything and returns the total number of bytes
      em.complete(Response.ok().header(HEADER_MAX_SIZE, max).build());
    });
  }
}

@QuarkusTest
class ApiTest {
  private static final Logger LOGGER = Logger.getLogger(ApiTest.class);
  @Inject
  ApiClient apiClient;
  @Inject
  ApiSecondClient apiSecondClient;

  @Test
  void testSendingProxy() {
    InputStream inputStream = new FakeInputStream(SIZE_1GB); // See code where FakeInputStream is written
    long size = apiClient.sendInputStream(inputStream);
    LOGGER.infof("Length %d", size);
  }

  @Test
  void testSendingDirect() {
    InputStream inputStream = new FakeInputStream(SIZE_1GB);
    long size = apiSecondClient.sendInputStream(inputStream);
    LOGGER.infof("Length:  %d", size);
  }
}
```
Test testSendingProxy is running but a subset of the inputStream only (around 1/2 MB), while testSendingDirect is OK.

So before doing anything else, it shall be ok at least for that too, right?

I don't have any idea yet.

fredericBregier commented:
Sorry for the spam alike...

I tried the following changes. I added a test with concurrent client (10 threads). With previous implementation based on VertxBlockingStream, it was always in error (length not ok). With this version, 8/10 threads are ok, still 2 not. So I hope I'm not so far from a correct answer.

No blocking but early closing of the InputStream... Still investigating

(Applied on all 3 versions with small adaptations to context)

```
  public static class VertxBlockingInput implements Handler<Buffer> {
    protected final HttpClientResponse request;
    protected final BlockingDeque<Buffer> inputOverflowX;
    private static final int INTERNAL_WAIT_MS = 20;
    protected boolean eof = false;
    protected Throwable readException;
    private final long timeout;
    private final int headerLen;

    public VertxBlockingInput(HttpClientResponse response, long timeout) {
      this.request = response;
      this.headerLen = getLengthFromHeader();
      this.timeout = timeout;
      response.pause();
      inputOverflowX = new LinkedBlockingDeque<>(Math.max(10, ConfigurationVertxInputStream.INITIAL_PREFETCH));
      response.handler(this);
      try {
        response.endHandler(new Handler<Void>() {
          @Override
          public void handle(Void event) {
            eof = true;
          }
        });
        response.exceptionHandler(new Handler<Throwable>() {
          @Override
          public void handle(Throwable event) {
            readException = new IOException(event);
            try {
              Buffer d = inputOverflowX.poll(INTERNAL_WAIT_MS, TimeUnit.MILLISECONDS);
              while (d != null) {
                d.getByteBuf().release();
                d = inputOverflowX.poll(INTERNAL_WAIT_MS, TimeUnit.MILLISECONDS);
              }
            } catch (InterruptedException e) {
              Thread.currentThread().interrupt();
            }
          }
        });
        response.fetch(ConfigurationVertxInputStream.INITIAL_PREFETCH);
      } catch (IllegalStateException e) {
        //already ended
        eof = true;
      }
    }

    protected ByteBuf readBlocking() throws IOException {
      long expire = System.currentTimeMillis() + timeout;
      Buffer ret;
      try {
        ret = inputOverflowX.poll(INTERNAL_WAIT_MS, TimeUnit.MILLISECONDS);
      } catch (InterruptedException e) {
        throw new InterruptedIOException(e.getMessage());
      }
      while (ret == null && !eof && readException == null) {
        long rem = expire - System.currentTimeMillis();
        if (rem <= 0) {
          //everything is broken, if read has timed out we can assume that the underling connection
          //is wrecked, so just close it
          request.netSocket().close();
          IOException throwable = new IOException("Read timed out");
          readException = throwable;
          throw throwable;
        }

        if (Context.isOnEventLoopThread()) {
          throw new BlockingNotAllowedException("Attempting a blocking read on io thread");
        }
        if (readException != null) {
          throw new IOException(readException);
        }
        try {
          ret = inputOverflowX.poll(INTERNAL_WAIT_MS, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
          throw new InterruptedIOException(e.getMessage());
        }
      }
      if (!eof) {
        request.fetch(1);
      }
      return ret == null ? null : ret.getByteBuf();
    }

    @Override
    public void handle(Buffer event) {
      long expire = System.currentTimeMillis() + timeout;
      while (readException == null) {
        long rem = expire - System.currentTimeMillis();
        if (rem <= 0) {
          //everything is broken, if read has timed out we can assume that the underling connection
          //is wrecked, so just close it
          request.netSocket().close();
          IOException throwable = new IOException("Write timed out");
          readException = throwable;
        }
        try {
          if (inputOverflowX.offerLast(event, INTERNAL_WAIT_MS, TimeUnit.MILLISECONDS)) {
            return;
          }
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
      }
    }

    public int readBytesAvailable() {
      Buffer buf = inputOverflowX.peek();
      if (buf != null) {
        int len = buf.getByteBuf().readableBytes();
        if (len > 0) {
          return len;
        }
      }
      return headerLen;
    }

    private int getLengthFromHeader() {
      String length = request.getHeader(HttpHeaders.CONTENT_LENGTH);
      if (length == null) {
        return 0;
      }
      try {
        return Integer.parseInt(length);
      } catch (NumberFormatException e) {
        Long.parseLong(length); // ignore the value as can only return an int anyway
        return Integer.MAX_VALUE;
      }
    }
  }
```

fredericBregier commented:
Stopping research for now: debugging gives me that it seems Http1xServerConnection.onEnd is called too early, such that there are missing chunks, but I don't get why.
Sending with 1 thread, even multiple times is OK. But when several threads in function, it turns wrong (6/10 in average),  using QuarkusRestClientBuilder or native Quarkus Rest client.

fredericBregier committed:
Improve VertxInputStream

In particular, improve VertxBlockingInput

fredericBregier commented:
@vietj I ended up with a new version which seems far more stable, both in sequential and concurrent usages.
- The VertBlockingInput is now far less blocking, only on necessary parts. 
- AsyncInputStream is inlining the dependence from Vertx (InboundBuffer) with light modification to simplify it

fredericBregier committed:
Add concurrent tests

fredericBregier commented:
@vietj Hi, I know the end and begiining of year is not the most favorable time to review ;-)
Any chance that this time my proposal is closer to your thought ?

vietj commented:
@fredericBregier I need to dedicate time to this, now seems more favorable indeed!

fredericBregier committed:
Fix javadoc

fredericBregier commented:
Any news? I guess next quarkus release is probably taken all available time...

fredericBregier commented:
Hi @vietj 
I just rebase the code such that to make it easier (no issue with rebase at all).

fredericBregier commented:
Hi @vietj 
Any news? Maybe you're thinking to another way (vertx native way?) or is this proposal not ok?

vietj commented:
@fredericBregier sorry no time for that, I am very overloaded with the recent releases we have done I will try to have a look soon (hint: this code is quite complex)

fredericBregier commented:
I understand. I know the recent releases were huge!! No problem

fredericBregier commented:
Hi @vietj 
Just a short notice on last Quarkus "main" version on ClientSendRequerstHandler with another implementation (far more simpler than mine): I've tried this implementation and it seems to fit. I've got some new errors in my tests but not related (headers issue). I need to investigate.
I think there is still some space for optimization on VertxXXXInputStream implementation, but not that mandatory yet.

fredericBregier commented:
@vietj OK, I found out that the next commit was fixing the header issue (I was on the commit level, so missing this one). I have to test however.

Do you now when those commits will be put in an official release ?

geoand commented:
Hi @fredericBregier 

Those commits are already part of Quarkus 3.8.3.

I am sorry that I had not realized my PR was essentially doing the same thing as yours.

geoand commented:
Glad to hear it works for you!

Improvements are absolutely welcome 😁

fredericBregier commented:
@geoand The commit 39282 is not in the 3.8.3 release which turns into bad header mapping (no header at all).
I believe a quick fix shall be done (3.8.4) since it breaks inputstream handling (I believe however this does not touch other cases)

geoand commented:
You are right, that was my fault for not properly labeling the PR.

It should be part of the next 3.8 release

fredericBregier commented:
No problem. As soon as I saw it, I sent this message in order to fix it in next release.
Thank you again !

geoand commented:
Thank you for raising the issue!
